cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -1222,25 +1243,28 @@ public void configure(Configuration conf, CConfiguration cConf,
       double minHeapRatio = cConf.getDouble(Configs.Keys.HEAP_RESERVED_MIN_RATIO);
 
       // Optionally overrides the settings from the runtime arguments
-      Map<String, String> configs = SystemArguments.getTwillContainerConfigs(runtimeArgs, resources.getMemoryMB());
+      Map<String, String> configs = SystemArguments.getTwillContainerConfigs(runtimeArgs, memoryMB);
       if (configs.containsKey(Configs.Keys.JAVA_RESERVED_MEMORY_MB)) {
         reservedMemory = Integer.parseInt(configs.get(Configs.Keys.JAVA_RESERVED_MEMORY_MB));
       }
       if (configs.containsKey(Configs.Keys.HEAP_RESERVED_MIN_RATIO)) {
         minHeapRatio = Double.parseDouble(configs.get(Configs.Keys.HEAP_RESERVED_MIN_RATIO));
       }
 
-      int maxHeapSize = org.apache.twill.internal.utils.Resources.computeMaxHeapSize(resources.getMemoryMB(),
+      int maxHeapSize = org.apache.twill.internal.utils.Resources.computeMaxHeapSize(memoryMB,
                                                                                      reservedMemory, minHeapRatio);
 
       String jvmOpts = cConf.get(Constants.AppFabric.PROGRAM_JVM_OPTS) + " -Xmx" + maxHeapSize + "m";
       // If the job conf has the jvm opts set programmatically, then prepend it with what being determined above;
       // otherwise, append. The reason is, we want to have the one set programmatically has higher precedence,
-      // while the one set cluster wide has lower precedence.
+      // while the one set cluster wide (in mapred-site.xml) has lower precedence.
       jvmOpts = isProgrammaticConfig(conf, javaOptsKey)
         ? jvmOpts + " " + conf.get(javaOptsKey, "")
         : conf.get(javaOptsKey, "") + " " + jvmOpts;
-      conf.set(javaOptsKey, jvmOpts);
+      conf.set(javaOptsKey, jvmOpts.trim());
+
+      LOG.debug("MapReduce {} task container memory is {} MB with maximum heap memory size of {} MB{}",
```

Comment:
I would prefer if we could log all these together with where they come from. I know it's a lot of logging in the code, but it would make developer's life a lot easier if they can see where, for example, the memory setting came from. It happens so easily that you have a preference at app level and don't realize it...